File: packages/ctf-core/src/optimizer.ts
/**
 * Auto-optimization analyzer for CTF format
 */

import {
  JsonValue,
  JsonObject,
  JsonArray,
  DataAnalysis,
  OptimizationStrategy,
  Delimiter,
} from './types.js';
import {
  isObject,
  isArray,
  haveSameStructure,
  allPrimitiveValues,
  countOccurrences,
  estimateTokens,
} from './utils.js';

export class Optimizer {
  /**
   * Analyze data structure and return recommendations
   */
  analyze(data: JsonValue): DataAnalysis {
    const analysis: DataAnalysis = {
      totalArrays: 0,
      tabularCandidates: 0,
      columnarCandidates: 0,
      repeatedValues: new Map(),
      estimatedTokens: 0,
      delimiterFrequency: new Map([
        ['|', 0],
        [',', 0],
        ['\t', 0],
      ]),
    };

    this.analyzeValue(data, analysis);

    // Estimate tokens from JSON representation
    analysis.estimatedTokens = estimateTokens(JSON.stringify(data));

    return analysis;
  }

  /**
   * Recursively analyze a value
   */
  private analyzeValue(value: JsonValue, analysis: DataAnalysis): void {
    if (isArray(value)) {
      analysis.totalArrays++;

      // Check if tabular candidate
      if (this.isTabularCandidate(value)) {
        analysis.tabularCandidates++;
      }

      // Check if columnar candidate
      if (this.isColumnarCandidate(value)) {
        analysis.columnarCandidates++;
      }

      // Analyze array elements
      for (const item of value) {
        this.analyzeValue(item, analysis);
      }

      // Count delimiter occurrences in array data
      this.analyzeDelimiterFrequency(value, analysis);
    } else if (isObject(value)) {
      for (const val of Object.values(value)) {
        this.analyzeValue(val, analysis);
      }
    } else if (typeof value === 'string') {
      // Count repeated values
      const count = analysis.repeatedValues.get(value) || 0;
      analysis.repeatedValues.set(value, count + 1);
    }
  }

  /**
   * Check if array is a tabular candidate
   */
  private isTabularCandidate(arr: JsonArray): boolean {
    return arr.length >= 3 && haveSameStructure(arr) && allPrimitiveValues(arr);
  }

  /**
   * Check if array is a columnar candidate
   */
  private isColumnarCandidate(arr: JsonArray): boolean {
    // Columnar makes sense for very large arrays with repetition
    if (arr.length < 1000) return false;
    if (!haveSameStructure(arr)) return false;
    if (!allPrimitiveValues(arr)) return false;

    // Check for high repetition in columns
    const objects = arr as JsonObject[];
    const keys = Object.keys(objects[0]);

    for (const key of keys) {
      const values = objects.map((obj) => String(obj[key]));
      const uniqueValues = new Set(values);

      // If any column has >30% repetition, it's a good candidate
      const repetitionRatio = 1 - uniqueValues.size / values.length;
      if (repetitionRatio > 0.3) {
        return true;
      }
    }

    return false;
  }

  /**
   * Analyze which delimiter appears least in the data
   */
  private analyzeDelimiterFrequency(arr: JsonArray, analysis: DataAnalysis): void {
    const jsonStr = JSON.stringify(arr);

    analysis.delimiterFrequency.set('|', (analysis.delimiterFrequency.get('|') || 0) + countOccurrences(jsonStr, '|'));
    analysis.delimiterFrequency.set(',', (analysis.delimiterFrequency.get(',') || 0) + countOccurrences(jsonStr, ','));
    analysis.delimiterFrequency.set('\t', (analysis.delimiterFrequency.get('\t') || 0) + countOccurrences(jsonStr, '\t'));
  }

  /**
   * Recommend optimization strategy based on analysis
   */
  recommendStrategy(analysis: DataAnalysis): OptimizationStrategy {
    if (analysis.columnarCandidates > 0) {
      return 'columnar';
    }

    if (analysis.totalArrays > 0) {
      const tabularRatio = analysis.tabularCandidates / analysis.totalArrays;
      if (tabularRatio > 0.5) {
        return 'tabular-heavy';
      }
    }

    return 'balanced';
  }

  /**
   * Choose optimal delimiter based on data
   */
  chooseDelimiter(analysis: DataAnalysis): Delimiter {
    const frequencies = analysis.delimiterFrequency;

    // Find delimiter with minimum occurrences
    let minDelimiter: Delimiter = '|';
    let minCount = Infinity;

    for (const [delimiter, count] of frequencies) {
      if (count < minCount) {
        minCount = count;
        minDelimiter = delimiter as Delimiter;
      }
    }

    // Prefer | > \t > , for token efficiency (all else being equal)
    if (frequencies.get('|') === frequencies.get('\t') && frequencies.get('|') === frequencies.get(',')) {
      return '|';
    }

    return minDelimiter;
  }

  /**
   * Estimate token savings from using references
   */
  estimateReferenceSavings(analysis: DataAnalysis, minOccurrences: number = 3, minLength: number = 5): number {
    let totalSavings = 0;

    for (const [value, count] of analysis.repeatedValues) {
      if (count >= minOccurrences && value.length >= minLength) {
        // Savings = (occurrences - 1) × length - (occurrences × 2)
        const savings = (count - 1) * value.length - count * 2;
        if (savings > 0) {
          totalSavings += savings;
        }
      }
    }

    return totalSavings;
  }
}

File: python/ctf/optimizer.py
"""Auto-optimization for CTF format."""

from typing import Any, Literal, Dict
from ctf.types import JsonValue


class Optimizer:
    """Analyzes data to select optimal encoding strategy."""

    def __init__(self, data: JsonValue) -> None:
        """
        Initialize optimizer with data.

        Args:
            data: JSON value to analyze
        """
        self.data = data
        self.stats = {
            "total_arrays": 0,
            "tabular_arrays": 0,
            "total_strings": 0,
            "repeated_strings": 0,
            "max_depth": 0,
            "total_objects": 0,
        }
        self._analyze(data, depth=0)

    def _analyze(self, value: Any, depth: int = 0) -> None:
        """
        Recursively analyze data structure.

        Args:
            value: Value to analyze
            depth: Current nesting depth
        """
        self.stats["max_depth"] = max(self.stats["max_depth"], depth)

        if isinstance(value, dict):
            self.stats["total_objects"] += 1
            for v in value.values():
                self._analyze(v, depth + 1)

        elif isinstance(value, list):
            self.stats["total_arrays"] += 1
            if self._is_tabular_array(value):
                self.stats["tabular_arrays"] += 1
            for item in value:
                self._analyze(item, depth + 1)

        elif isinstance(value, str):
            self.stats["total_strings"] += 1

    def _is_tabular_array(self, arr: list[Any]) -> bool:
        """
        Check if array is suitable for tabular encoding.

        Args:
            arr: Array to check

        Returns:
            True if array should use tabular encoding
        """
        if len(arr) < 3:
            return False

        # Check if all items are objects with same keys
        if not all(isinstance(item, dict) for item in arr):
            return False

        if not arr:
            return False

        first_keys = set(arr[0].keys())
        return all(set(item.keys()) == first_keys for item in arr)

    def select_delimiter(self, data: Any) -> Literal[",", "|", "\t"]:
        """
        Select optimal delimiter for tabular arrays.

        Args:
            data: Data to analyze

        Returns:
            Best delimiter character
        """
        # Count occurrences of potential delimiters in string values
        delimiter_counts = {"|": 0, ",": 0, "\t": 0}

        def count_in_strings(value: Any) -> None:
            if isinstance(value, str):
                delimiter_counts["|"] += value.count("|")
                delimiter_counts[","] += value.count(",")
                delimiter_counts["\t"] += value.count("\t")
            elif isinstance(value, dict):
                for v in value.values():
                    count_in_strings(v)
            elif isinstance(value, list):
                for item in value:
                    count_in_strings(item)

        count_in_strings(data)

        # Choose delimiter that appears least in data
        return min(delimiter_counts.items(), key=lambda x: x[1])[0]

    def should_use_references(self) -> bool:
        """
        Determine if reference compression should be used.

        Returns:
            True if references would be beneficial
        """
        # Use references if there are repeated strings and decent depth
        return self.stats["total_strings"] > 10 and self.stats["max_depth"] > 2

    def should_use_columnar(self, array_length: int) -> bool:
        """
        Determine if columnar encoding should be used for an array.

        Args:
            array_length: Length of the array

        Returns:
            True if columnar encoding would be beneficial
        """
        # Use columnar for very large arrays (1000+ rows)
        return array_length >= 1000

    def get_recommendations(self) -> Dict[str, Any]:
        """
        Get optimization recommendations.

        Returns:
            Dictionary of recommendations
        """
        return {
            "delimiter": self.select_delimiter(self.data),
            "use_references": self.should_use_references(),
            "tabular_arrays": self.stats["tabular_arrays"],
            "total_arrays": self.stats["total_arrays"],
            "max_depth": self.stats["max_depth"],
        }
